### PR TITLE
Implement support for mapping between long double and Float80

### PR DESCRIPTION
**Pull request by stephentyrone**
This needs more tests, but superficially it seems to work; I'd like to get some review and make sure I'm not missing something silly.

Also exposes all the tgmath wrappers on Float80.

Resolves:
[SR-2046](https://bugs.swift.org/browse/SR-2046)
<rdar://problem/27196587>
<rdar://problem/30453167>

### CODE REVIEW

**Review by moiseev — APPROVED** [review]
stdlib changes LGTM

File: stdlib/public/core/CTypes.swift [hunk 1]
```diff
@@ -62,7 +62,12 @@ public typealias CFloat = Float
 /// The C 'double' type.
 public typealias CDouble = Double
 
-// FIXME: long double
+/// The C 'long double' type.
+#if arch(x86_64) || arch(i386)
+public typealias CLongDouble = Float80
+#else
+public typealias CLongDouble = Double
```
Comment (jrose-apple) [inline stdlib/public/core/CTypes.swift]: Should this be more explicit to rule out architectures we haven't thought about? Technically there's support for pp64le in tree.

**Review by jrose-apple — COMMENTED** [review]
I forgot we already had a Float80 in tree and mostly we just need to turn it on. @rjmccall, can you think of anything that would trip us up with regards to size or alignment? (Or anything else.)

File: stdlib/public/core/CTypes.swift (hunk 1 above)
Comment (stephentyrone) [reply stdlib/public/core/CTypes.swift]: Yeah, I think ppc64le uses a 128b long double, as does non-Darwin AArch64, IIRC.

Comment (jrose-apple) [reply stdlib/public/core/CTypes.swift]: I'm not saying you need to implement those now, but the typealias shouldn't lie.

**Comment by stephentyrone** [comment]
@jrose-apple @rjmccall Unless I missed something, I don't think there are any gotchas w.r.t size / alignment: 16B for both in both C and Swift.

**Comment by stephentyrone** [comment]
@swift-ci Please test.

File: stdlib/public/core/CTypes.swift [hunk 2]
```diff
@@ -63,11 +63,26 @@ public typealias CFloat = Float
...
 #else
 public typealias CLongDouble = Double
 #endif
+#elseif os(Windows)
+// On Windows, long double is always Double.
+public typealias CLongDouble = Double
+#elseif os(Linux)
+// On Linux/x86, long double is Float80.
+// TODO: Fill in definitions for additional architectures as needed. IIRC
+// armv7 should map to Double, but arm64 and ppc64le should map to Float128,
```
Comment (compnerd) [inline stdlib/public/core/CTypes.swift]: This is correct, I would drop the "IIRC".  ARMv7 Linux maps `long double` to `double` both of which are FP64.  ARM64 and PPC64 both use FP128.

Comment (jrose-apple) [reply stdlib/public/core/CTypes.swift]: Hey, @compnerd, want to implement FP128? ;-)

Comment (compnerd) [reply stdlib/public/core/CTypes.swift]: I might ... 🤦‍♂️

Comment (stephentyrone) [reply stdlib/public/core/CTypes.swift]: We all *want* to implement `Float128`, it's just a matter of free time =)

Comment (jrose-apple) [reply stdlib/public/core/CTypes.swift]: I'm pretty sure it's not going to be harder than Float80. It's all just forwarding to LLVM.

…well, okay, the fact that tgmath doesn't have overloads for it is kind of annoying.

**Pull request closed without merging**